File: java/SearchInsertDelete.java
import java.util.Random;
import java.util.concurrent.Semaphore;
import java.util.*;

//Solution taken from little book of semaphores and applied to java
public class SearchInsertDelete {

    static Semaphore insertMutex = new Semaphore(1);
    static Semaphore noSearcher = new Semaphore(1);
    static Semaphore noInserter = new Semaphore(1);
    static Lightswitch searchSwitch = new Lightswitch();
    static Lightswitch insertSwitch = new Lightswitch();
    private static List<Integer> myList;

    static int maxSearchers = 50;
    static int maxInserters = 20;
    static int maxDeleters = 10;

    public static class Lightswitch {
        private Semaphore mutex = new Semaphore(1);
        private int counter = 0;

        public void wait(Semaphore semaphore) {
            try {
                mutex.acquire();
                counter++;
                if (counter == 1) {
                    semaphore.acquire();
                }
                mutex.release();
            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }

        public void signal(Semaphore semaphore) {
            try {
                mutex.acquire();
                counter--;
                if (counter == 0) {
                    semaphore.release();
                }
                mutex.release();
            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }
    }

    public static void sleepThread(int time) {
        try {
            Thread.sleep((long) (Math.random() * time));
        } catch (InterruptedException e) {
            System.out.println(e.getMessage());
        }
    }

    public static void main(String args[]) {
        myList = new ArrayList<Integer>();

        System.out.println("Starting inserters");
        myList.add(1);
        myList.add(2);
        myList.add(3);

        Inserter inserter = new Inserter();
        Searcher searcher = new Searcher();
        Deleter deleter = new Deleter();

        // init inserters
        Thread[] inserters = new Thread[maxInserters];
        for (int i = 0; i < inserters.length; i++) {
            inserters[i] = new Thread(inserter);
            inserters[i].setName("inserter " + i);
            inserters[i].start();
        }

        System.out.println("Starting searchers");
        // init searchers
        Thread[] searchers = new Thread[maxSearchers];
        for (int i = 0; i < searchers.length; i++) {
            searchers[i] = new Thread(searcher);
            searchers[i].setName("searcher " + i);
            searchers[i].start();
        }

        System.out.println("Starting deleters");
        // init deleters
        Thread[] deleters = new Thread[maxDeleters];
        for (int i = 0; i < deleters.length; i++) {
            deleters[i] = new Thread(deleter);
            deleters[i].setName("deleter " + i);
            deleters[i].start();
        }
    }

    static class Inserter implements Runnable {
        @Override
        public void run() {
            try {
                sleepThread(250);
                insertSwitch.wait(noInserter);
                insertMutex.acquire();
                Random rand = new Random();
                int val = rand.nextInt(100);
                System.out.println("Inserting: " + val);
                myList.add(val);
                sleepThread(250);
                insertMutex.release();
                insertSwitch.signal(noInserter);
            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }
        }
    }

    static class Searcher implements Runnable {
        @Override
        public void run() {
            sleepThread(250);
            searchSwitch.wait(noSearcher);
            System.out.println("Searched: " + myList.get(0));
            sleepThread(250);
            searchSwitch.signal(noSearcher);
        }
    }

    static class Deleter implements Runnable {
        @Override
        public void run() {
            try {
                sleepThread(250);
                noSearcher.acquire();
                noInserter.acquire();
                int val = myList.get(myList.size() - 1);
                System.out.println("Deleting " + val);
                myList.remove(myList.size() - 1);
                sleepThread(250);
                noInserter.release();
                noSearcher.release();
            } catch (InterruptedException e) {
                System.out.println(e.getMessage());
            }

        }
    }
}

File: golang/search_insert_delete.go
package main

import (
	"container/list"
	"fmt"
	"math/rand"
	"sync"
	"time"

	"github.com/pkg/profile"
)

var mutexDelete sync.RWMutex
var mutexInsert sync.Mutex
var wg sync.WaitGroup
var myList list.List

var searchers = 50
var inserters = 20
var deleters = 10

func sleepThread(counter int) {
	time.Sleep(time.Duration(rand.Intn(counter)) * time.Millisecond)
}
func main() {
	defer profile.Start().Stop()
	fmt.Println("Starting inserters")
	myList.PushBack(1)
	myList.PushBack(2)
	myList.PushBack(3)

	wg.Add(inserters)
	go func() {
		for i := 0; i < inserters; i++ {
			sleepThread(250)
			go insert()
		}
	}()
	fmt.Println("Starting searchers")
	wg.Add(searchers)
	go func() {
		for i := 0; i < searchers; i++ {
			sleepThread(100)
			go search()
		}
	}()

	fmt.Println("Starting deleters")
	wg.Add(deleters)
	go func() {
		for i := 0; i < searchers; i++ {
			sleepThread(500)
			go delete()
		}
	}()
	wg.Wait()
}

func search() {
	sleepThread(250)
	//search only needs to readlock the delete mutex
	mutexDelete.RLock()
	fmt.Println("Searched: ", myList.Front().Value)
	sleepThread(250)
	mutexDelete.RUnlock()
	wg.Done()
}

func insert() {
	sleepThread(250)
	//Insert needs to lock the insert mutex to make sure no other inserts and also readlock the delete mutex
	mutexInsert.Lock()
	mutexDelete.RLock()
	val := rand.Intn(100)
	fmt.Printf("Inserting: %d\n", val)
	myList.PushFront(val)
	sleepThread(250)
	mutexInsert.Unlock()
	mutexDelete.RUnlock()
	wg.Done()
}

func delete() {
	sleepThread(250)
	//Need to lock both the insert and delete mutex so no one else is doing anything
	mutexInsert.Lock()
	mutexDelete.Lock()
	val := myList.Back()
	fmt.Printf("Deleting: %d\n", val.Value)
	myList.Remove(val)
	sleepThread(250)
	mutexInsert.Unlock()
	mutexDelete.Unlock()
	wg.Done()
}
